From 4f6c1dc3c48a1b2fa7c06206e6366bcfaa33f3f7 Mon Sep 17 00:00:00 2001
From: Christian Breunig <christian@breunig.cc>
Date: Fri, 22 Mar 2024 11:33:27 +0000
Subject: [PATCH] ixgbe: always enable support for unsupported SFP+ modules

---
 ixgbe_param.c | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/ixgbe_param.c b/ixgbe_param.c
index 71197b7..dac33ca 100644
--- a/ixgbe_param.c
+++ b/ixgbe_param.c
@@ -307,7 +307,7 @@ IXGBE_PARAM(LRO, "Large Receive Offload (0,1), default 0 = off");
  * Default Value: 0
  */
 IXGBE_PARAM(allow_unsupported_sfp, "Allow unsupported and untested "
-	    "SFP+ modules on 82599 based adapters, default 0 = Disable");
+	    "SFP+ modules on 82599 based adapters, default 1 = Enable");
 
 /* Enable/disable support for DMA coalescing
  *
@@ -1133,8 +1133,8 @@ void ixgbe_check_options(struct ixgbe_adapter *adapter)
 	struct ixgbe_option opt = {
 			.type = enable_option,
 			.name = "allow_unsupported_sfp",
-			.err  = "defaulting to Disabled",
-			.def  = OPTION_DISABLED
+			.err  = "defaulting to Enabled",
+			.def  = OPTION_ENABLED
 		};
 #ifdef module_param_array
 		if (num_allow_unsupported_sfp > bd) {
@@ -1150,7 +1150,11 @@ void ixgbe_check_options(struct ixgbe_adapter *adapter)
 			}
 #ifdef module_param_array
 		} else {
+			if (opt.def == OPTION_ENABLED) {
+				adapter->hw.allow_unsupported_sfp = true;
+			} else {
 				adapter->hw.allow_unsupported_sfp = false;
+			}
 		}
 #endif
 	}
-- 
2.39.2

