From 23f4e8789b1bdcc0442b6d57216e5184c1bd97c8 Mon Sep 17 00:00:00 2001
From: Christian Poessinger <christian@poessinger.com>
Date: Mon, 17 Jan 2022 06:43:26 +0000
Subject: [PATCH] Enable PAM support

---
 debian/rules      | 2 +-
 default_options.h | 4 ++--
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/debian/rules b/debian/rules
index 3e944d8..3a9f7a7 100755
--- a/debian/rules
+++ b/debian/rules
@@ -28,7 +28,7 @@ override_dh_installinit:
	dh_installinit -R --name dropbear

 override_dh_auto_configure:
-	dh_auto_configure -- --disable-bundled-libtom \
+	dh_auto_configure -- --disable-bundled-libtom --enable-pam \
	  CC='$(CC)' CFLAGS='$(CFLAGS)' $(CONFFLAGS)

 override_dh_install:
diff --git a/default_options.h b/default_options.h
index 9000fcc..0db7366 100644
--- a/default_options.h
+++ b/default_options.h
@@ -179,7 +179,7 @@ group1 in Dropbear server too */

 /* Authentication Types - at least one required.
    RFC Draft requires pubkey auth, and recommends password */
-#define DROPBEAR_SVR_PASSWORD_AUTH 1
+#define DROPBEAR_SVR_PASSWORD_AUTH 0

 /* Note: PAM auth is quite simple and only works for PAM modules which just do
  * a simple "Login: " "Password: " (you can edit the strings in svr-authpam.c).
@@ -187,7 +187,7 @@ group1 in Dropbear server too */
  * but there's an interface via a PAM module. It won't work for more complex
  * PAM challenge/response.
  * You can't enable both PASSWORD and PAM. */
-#define DROPBEAR_SVR_PAM_AUTH 0
+#define DROPBEAR_SVR_PAM_AUTH 1

 /* ~/.ssh/authorized_keys authentication */
 #define DROPBEAR_SVR_PUBKEY_AUTH 1
--
2.20.1
