From 20ff690dcf8ab7e800e1dc0c4fd83041d039fb3d Mon Sep 17 00:00:00 2001
From: Daniil Baturin <daniil@baturin.org>
Date: Wed, 6 Nov 2024 15:58:10 +0000
Subject: [PATCH] Enable PCRE2 in Debian package builds

---
 debian/rules | 1 +
 1 file changed, 1 insertion(+)

diff --git a/debian/rules b/debian/rules
index 21094c62d9..ea248ff528 100755
--- a/debian/rules
+++ b/debian/rules
@@ -89,6 +89,7 @@ override_dh_auto_configure:
 		--enable-vty-group=frrvty \
 		--enable-configfile-mask=0640 \
 		--enable-logfile-mask=0640 \
+		--enable-pcre2posix \
 		# end
 
 override_dh_auto_install:
-- 
2.39.5

